sql/resource_query.go

```diff
@@ -0,0 +1,169 @@
+	// 	"parameter.0.text_value",
+	// 	"parameter.0.numeric_value",
+	// 	"parameter.0.enum_value",
+	// }
+	// for _, f := range valuesAlof {
+	// 	m.SchemaPath("parameter", strings.TrimPrefix(f, "parameter.0.")).SetAtLeastOneOf(valuesAlof)
+	// }
+	return m
+}
+
+type queryCreateStruct struct {
+	sql.CreateQueryRequestQuery
+}
+
+func (queryCreateStruct) Aliases() map[string]map[string]string {
+	return map[string]map[string]string{
+		"sql.queryCreateStruct": queryAliasMap,
+	}
+}
+
+func (queryCreateStruct) CustomizeSchema(s *common.CustomizableSchema) *common.CustomizableSchema {
+	return s
+}
+
+type queryUpdateStruct struct {
+	sql.UpdateQueryRequestQuery
+}
+
+func (queryUpdateStruct) Aliases() map[string]map[string]string {
+	return map[string]map[string]string{
+		"sql.queryUpdateStruct": queryAliasMap,
+	}
+}
+
+func (queryUpdateStruct) CustomizeSchema(s *common.CustomizableSchema) *common.CustomizableSchema {
+	return s
+}
+
+func ResourceQuery() common.Resource {
+	s := common.StructToSchema(queryStruct{}, nil)
+	return common.Resource{
+		Create: func(ctx context.Context, d *schema.ResourceData, c *common.DatabricksClient) error {
+			w, err := c.WorkspaceClient()
+			if err != nil {
+				return err
+			}
+			var q queryCreateStruct
+			common.DataToStructPointer(d, s, &q)
+			apiQuery, err := w.Queries.Create(ctx, sql.CreateQueryRequest{
+				Query: &q.CreateQueryRequestQuery,
+			})
+			if err != nil {
+				return err
+			}
+			d.SetId(apiQuery.Id)
+			owner := d.Get("owner_user_name").(string)
+			if owner != "" {
+				_, err = w.Queries.Update(ctx, sql.UpdateQueryRequest{
+					Query: &sql.UpdateQueryRequestQuery{
+						OwnerUserName: owner,
+					},
+					Id:         apiQuery.Id,
+					UpdateMask: "owner_user_name",
+				})
+			}
+			return err
+		},
+		Read: func(ctx context.Context, d *schema.ResourceData, c *common.DatabricksClient) error {
+			w, err := c.WorkspaceClient()
+			if err != nil {
+				return err
+			}
+			apiQuery, err := w.Queries.GetById(ctx, d.Id())
+			if err != nil {
+				log.Printf("[WARN] error getting query by ID: %v", err)
+				return err
+			}
+			parentPath := d.Get("parent_path").(string)
+			if parentPath != "" && strings.HasPrefix(apiQuery.ParentPath, "/Workspace") && !strings.HasPrefix(parentPath, "/Workspace") {
+				apiQuery.ParentPath = strings.TrimPrefix(parentPath, "/Workspace")
+			}
+			return common.StructToData(queryStruct{Query: *apiQuery}, s, d)
+		},
+		Update: func(ctx context.Context, d *schema.ResourceData, c *common.DatabricksClient) error {
+			w, err := c.WorkspaceClient()
+			if err != nil {
+				return err
+			}
+			var q queryUpdateStruct
+			common.DataToStructPointer(d, s, &q)
+			updateMask := "display_name,query_text,warehouse_id,parameters"
+			for _, f := range []string{"run_as_mode", "owner_user_name", "description", "tags",
+				"apply_auto_limit", "catalog", "schema"} {
+				if d.HasChange(f) {
+					updateMask += "," + f
+				}
+			}
```

Comment:
For Terraform, because DataToStructPointer has values for all fields, can we just use `*` for updateMask?

Reply:
API spec says:
> To specify multiple fields in the field mask, use a comma as the separator (no space).

I'm not sure that `*` is supported.

Especially for `owner_user_name` if we don't have change, it will decrease a pressure on backend, as changing the owner is heavy operation (at least as I understood from the explanation of the corresponding team).